soc: arm: add FocalTech FT9001 SoC and evaluation board support

robert-cheng-1127 opened the pull request:
This PR adds support for the FT9001 fingerprint controller from FocalTech, which is specially designed for fingerprint sensor applications.

This PR includes initial support for the FT9001 evaluation board, with drivers as follows:

- soc infrastructure
- uart (register-based implementation)
- clock control
- reset
- device tree bindings

Note: Current implementation uses direct register operations as the FocalTech HAL is not yet integrated into modules/hal.

nordicjm reviewed (changes requested):
split into multiple comments before this gets reviewed e.g. one for soc, one for board

robert-cheng-1127 commented on soc/focaltech/ft9001/Kconfig.soc:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FT9001
+  bool "FocalTech FT9001 MCU"
```
Sorry for the oversight. I’ll review and fix all Kconfig files in this PR.

nordicjm replied:
this is not fixed https://github.com/zephyrproject-rtos/zephyr/pull/95959/commits/e54fa195398cddc5014d0e71b5d7b5835b17e87c

robert-cheng-1127 commented on drivers/serial/uart_ft9001.c:
```diff
@@ -0,0 +1,593 @@
+static void uart_ft9001_hw_init(UART_TypeDef *base, const struct uart_config *cfg,
+				uint32_t sys_freq)
+{
+	uint16_t bandrate_I;
```
Thanks for the review. 
The external HAL module isn’t ready yet, so this revision uses register-level access.
I’ll keep all driver-local identifiers in snake case.

nordicjm commented on boards/focaltech/ft9001_eval/ft9001_eval_defconfig:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2025, FocalTech Systems CO.,Ltd
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_FLASH_LOAD_OFFSET=0x1000
```
Put it into dts as a partition and set the `CONFIG_USE_DT_CODE_PARTITION` Kconfig

robert-cheng-1127 commented on soc/focaltech/ft9001/ft9001_core.h:
```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2025, FocalTech Systems CO.,Ltd
```
Hi @gmarull, 
sorry for the late reply.
I’ve trimmed ft9001_core.h to minimal CMSIS glue as a temporary shim. Plan is to remove it and switch to HAL sources. Is that acceptable?

nordicjm commented on soc/focaltech/ft9001/ft9001_core.h:
```diff
@@ -0,0 +1,129 @@
+typedef enum
+{
+/******  Cortex-M4 Processor Exceptions Numbers ****************************************************************/
+  NonMaskableInt_IRQn         = -14,    /*!< 2 Non Maskable Interrupt */
```
this is not fixed, I'm going to abandon this review, can you fix the issues everyone has commented, update the existing commits with the fixes, force push then ping people when the issues have been addressed and is it ready to review?
And before doing that, please read and ensure code complies with the coding standards available at:
https://docs.zephyrproject.org/latest/contribute/style/code.html
https://docs.zephyrproject.org/latest/contribute/style/cmake.html
https://docs.zephyrproject.org/latest/contribute/style/devicetree.html
https://docs.zephyrproject.org/latest/contribute/style/kconfig.html
And also the SoC/board porting guides available at:
https://docs.zephyrproject.org/latest/hardware/porting/soc_porting.html
https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html

- Thanks!

robert-cheng-1127 committed:
soc: focaltech: ft9001: add SoC support

Add initial support for the FocalTech FT9001 ARM Cortex-M SoC including:
- SoC Kconfig and dtsi integration points
- Early init and core setup
- Pinctrl base definitions

This prepares the tree for the FT9001 board and drivers.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
dts: arm: focaltech: add ft9001 SoC dtsi

Introduce the base devicetree description for the FT9001 SoC.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
dt-bindings: focaltech: ft9001: add clock, reset, uart

Add devicetree bindings for the FocalTech FT9001 SoC:
- Clock controller
- Reset controller
- UART

These bindings are required by the SoC and drivers added in subsequent
commits. Headers are placed under include/zephyr/dt-bindings/.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
boards: focaltech: ft9001_eval: add evaluation board

Add support for the FocalTech FT9001 evaluation board:
- Board dts/dtsi and pinmux
- defconfig and board.cmake
- Basic documentation

The board has been tested with the following Zephyr sample:
– samples/hello_world

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
drivers: reset: focaltech: add ft9001 reset controller driver

Add reset controller driver for the FocalTech FT9001 SoC.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
drivers: serial: focaltech: add ft9001 serial driver

Add the UART driver for the FocalTech FT9001 SoC.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 committed:
drivers: clock_control: focaltech: add ft9001 clock control driver

Add clock control driver for the FocalTech FT9001 SoC.

Signed-off-by: Robert Cheng <robert.cheng@focaltech-electronics.com>

robert-cheng-1127 commented:
Hi @kartben - thanks again for running the tests earlier.
I’ve addressed the issues reported by Twister, Documentation, Compliance, and Scancode.
Could you please re-run the checks when you have a moment?
Much appreciated!

nordicjm commented on soc/focaltech/ft9001/ft9001_xip.c:
```diff
@@ -0,0 +1,166 @@
+ * This routine must run from RAM (.ramfunc) because it erases the very first
+ * sector at 0x1000 0000 while XIP is active.
+ */
+__ramfunc void xip_return_to_boot(void)
```
why is this here?

robert-cheng-1127 replied:
Thanks for catching this. it's not XIP-related. I’ve removed the non-XIP path.

robert-cheng-1127 commented on drivers/serial/uart_ft9001.c:
```diff
@@ -0,0 +1,599 @@
+
+/* FT9001 UART register structure (based on HAL) */
+struct ft9001_uart_regs {
+	__IO uint8_t SCIBDL;     /* 0x00 - Baud Rate Low */
```
Apologies. corrected to snake case throughout and reverified.

nordicjm reviewed (commented):
PR is OK but commits need reordering, you can't add a board in a commit then in a later commit add the files that the board depends on, they must be added first

fabiobaltieri commented on drivers/serial/uart_ft9001.c:
```diff
@@ -0,0 +1,606 @@
+static int uart_ft9001_fifo_fill(const struct device *dev, const uint8_t *tx_data, int size)
+{
+	const struct uart_ft9001_config *config = DEV_CFG(dev);
+	int num_tx = 0;
```
I think you can drop the initializer since you have it in the for

gmarull commented on soc/focaltech/ft9001/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	default y
+
+config XIP
+	default y
```
already default?

robert-cheng-1127 commented:
@fabiobaltieri @nordicjm 
Could you do another review when you have some time? Thank you!

nordicjm reviewed (dismissed):
SoC/boards files OK

robert-cheng-1127 commented on boards/focaltech/ft9001_eval/doc/index.rst:
```diff
@@ -0,0 +1,44 @@
+
+.. _focaltech_ft9001_eval:
+
+.. zephyr:board:: ft9001_eval
```
Not yet, I don’t have a suitable photo right now.
I'll take one and upload it later today.

robert-cheng-1127 commented:
@nordicjm @gmarull @decsny 
If you wouldn’t mind, could you please take another look when you have a moment? Thanks!

fabiobaltieri commented on boards/focaltech/ft9001_eval/doc/ft9001_eval.webp:
plenty good, board images should be just for reference really, like giving you a nice visual confirmation that the board is the one you have in your hand, if you need actual information for it the best thing is probably to link a documentation on the vendo website or something, anyway it's ok now thanks

fabiobaltieri reviewed (commented):
just one more comment on the board and uart init, ci fail is unrelated it should pass now

robert-cheng-1127 commented on boards/focaltech/ft9001_eval/ft9001_eval.dts:
```diff
@@ -0,0 +1,54 @@
+	parity = "none";
+	stop-bits = "1";
+	data-bits = <8>;
+	hw-flow-control;
```
I'll take a look at that file. I'll update the PR once it's done.
